xds/src/main/java/io/grpc/xds/GrpcAuthorizationEngine.java

```diff
@@ -0,0 +1,381 @@
+/*
+ * Copyright 2021 The gRPC Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.grpc.xds;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+import com.google.auto.value.AutoValue;
+import io.grpc.Grpc;
+import io.grpc.Metadata;
+import io.grpc.ServerCall;
+import io.grpc.xds.internal.Matchers;
+import io.grpc.xds.internal.Matchers.CidrMatcher;
+import io.grpc.xds.internal.Matchers.StringMatcher;
+import java.net.InetAddress;
+import java.net.InetSocketAddress;
+import java.net.SocketAddress;
+import java.security.cert.Certificate;
+import java.security.cert.CertificateParsingException;
+import java.security.cert.X509Certificate;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import javax.annotation.Nullable;
+import javax.net.ssl.SSLPeerUnverifiedException;
+import javax.net.ssl.SSLSession;
+
+/**
+ * Implementation of gRPC server access control based on envoy RBAC protocol:
+ * https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/rbac/v3/rbac.proto
+ *
+ * <p>One GrpcAuthorizationEngine is initialized with one action type and a list of policies.
+ * Policies are examined sequentially in order in an any match fashion, and the first matched policy
+ * will be returned. If not matched at all, the opposite action type is returned as a result.
+ */
+public final class GrpcAuthorizationEngine {
+  private static final Logger log = Logger.getLogger(GrpcAuthorizationEngine.class.getName());
+  private final AuthConfig authConfig;
+
+  enum Action {
+    ALLOW,
+    DENY,
+  }
+
+  /** An authorization decision provides information about the decision type and the policy name
+   * identifier based on the authorization engine evaluation. */
+  @AutoValue
+  abstract static class AuthDecision {
```

Comment:
Looks like the whole rbac engine under `/internal` is the best? Do we have timeline when `"SDK Authz API"` will be introduced? Now I still put under `io.grpc.xds`, we can refactor to `internal`, because I was planning to remove the `CEL` thing later and we can do the move altogether.

Reply:
Yeah, I expect putting it in internal is best. I expect the other API in early Q3. We probably shouldn't do the delete and the move in the same commit, as that will make it harder to follow in the git history.

Reply:
k, can we do the move later when it is more clear where http filter in xds server would use it.